libsolidity/interface/CompilerStack.cpp

```diff
@@ -1205,13 +1206,76 @@ Json CompilerStack::ethdebug(Contract const& _contract, bool _runtime) const
 	solAssert(m_stackState >= AnalysisSuccessful, "Analysis was not successful.");
 	solAssert(_contract.contract);
 	solUnimplementedAssert(!isExperimentalSolidity());
+	Json result = Json::object();
+	Json contract = Json::object();
+	contract["name"] = _contract.contract->name();
+	Json definition = Json::object();
+	definition["sources"]["id"] = static_cast<int>(sourceIndices()[_contract.contract->sourceUnitName()]);
+	contract["definition"] = definition;
+	result["contract"] = contract;
+	result["environment"] = _runtime ? "call" : "create";
+	result["instructions"] = ethdebugInstructions(_contract, _runtime);
+	return result;
+}
+
+Json CompilerStack::ethdebugInstructions(Contract const& _contract, bool _runtime) const
+{
+	evmasm::LinkerObject const* object;
+	evmasm::Assembly const* assembly;
 	if (_runtime)
 	{
-		Json result = Json::object();
-		return result;
+		object = &_contract.runtimeObject;
+		assembly = _contract.evmRuntimeAssembly.get();
 	}
-	Json result = Json::object();
-	return result;
+	else
+	{
+		object = &_contract.object;
+		assembly = _contract.evmAssembly.get();
+	}
+
+	Json instructions = Json::array();
+	for (size_t i = 0; i < object->offsets.size(); ++i)
+	{
+		size_t offset = object->offsets[i];
+		size_t nextOffset = offset;
+		if (i + 1 < object->offsets.size())
+			nextOffset = object->offsets[i + 1];
+		Json instruction = Json::object();
+		instruction["offset"] = offset;
+		instruction["mnemonic"] = instructionInfo(static_cast<evmasm::Instruction>(object->bytecode[offset]), m_evmVersion).name;
+		bytes argumentData;
+		for (size_t args = offset + 1; args < nextOffset; ++args)
+			argumentData.emplace_back(object->bytecode[args]);
+		if (!argumentData.empty())
+		{
+			Json arguments = Json::array();
+			arguments.emplace_back("0x" + util::toHex(argumentData));
+			instruction["arguments"] = arguments;
+		}
+
+		solAssert(assembly->codeSections().size() == 1);
+		SourceLocation const& location = assembly->codeSections().at(0).items.at(i).location();
+		Json instructionContext = Json::object();
+		Json source = Json::object();
+		source["id"] = location.sourceName ? static_cast<int>(sourceIndices()[*location.sourceName]) : -1;
+		Json range = Json::object();
+		range["offset"] = location.start;
+		range["length"] = location.end;
+		// if (location.sourceName && m_sources.count(*location.sourceName) && location.start > -1 && location.end > -1)
+		// {
+		// 	std::string value = std::string{m_sources.at(*location.sourceName).charStream->text(location)};
+		// 	std::regex whitespace_re("\\s+");
+		// 	range["value"] = std::regex_replace(value, whitespace_re, " ");
+		// }
```

Comment:
this could also use a simple for loop over the string (aside from being commented out right now)

Reply:
this "value" thing reflecting the current source range as a string is not part of ethdebug, I just had it to see the actual values - I just remove those now